solidity/contracts/relay/Relay.sol

```diff
@@ -0,0 +1,507 @@
+    /// @param genesisProofLength The number of blocks required to accept a
+    /// proof.
+    function genesis(
+        bytes calldata genesisHeader,
+        uint256 genesisHeight,
+        uint64 genesisProofLength
+    ) external onlyOwner {
+        require(!ready, "Genesis already performed");
+
+        require(genesisHeader.length == 80, "Invalid genesis header length");
+
+        require(
+            genesisHeight % 2016 == 0,
+            "Invalid height of relay genesis block"
+        );
+
+        require(genesisProofLength < 2016, "Proof length excessive");
+        require(genesisProofLength > 0, "Proof length may not be zero");
+
+        genesisEpoch = uint64(genesisHeight / 2016);
+        currentEpoch = genesisEpoch;
+        uint256 genesisTarget = genesisHeader.extractTarget();
+        uint256 genesisTimestamp = genesisHeader.extractTimestamp();
+        epochs[genesisEpoch] = Epoch(
+            uint32(genesisTimestamp),
+            uint224(genesisTarget)
+        );
+        proofLength = genesisProofLength;
+        currentEpochDifficulty = BTCUtils.calculateDifficulty(genesisTarget);
+        ready = true;
+
+        emit Genesis(genesisHeight);
+    }
+
+    /// @notice Set the number of blocks required to accept a header chain.
+    /// @param newLength The required number of blocks. Must be less than 2016.
+    /// @dev For production, a high number (e.g. 20-50) is recommended.
+    /// Small numbers are accepted but should only be used for testing.
+    function setProofLength(uint64 newLength) external relayActive onlyOwner {
+        require(newLength < 2016, "Proof length excessive");
+        require(newLength > 0, "Proof length may not be zero");
+        require(newLength != proofLength, "Proof length unchanged");
+        proofLength = newLength;
+        emit ProofLengthChanged(newLength);
+    }
+
+    /// @notice Add a new epoch to the relay by providing a proof
+    /// of the difficulty before and after the retarget.
+    /// @param headers A chain of headers including the last X blocks before
+    /// the retarget, followed by the first X blocks after the retarget,
+    /// where X equals the current proof length.
+    /// @dev Checks that the first X blocks are valid in the most recent epoch,
+    /// that the difficulty of the new epoch is calculated correctly according
+    /// to the block timestamps, and that the next X blocks would be valid in
+    /// the new epoch.
+    /// We have no information of block heights, so we cannot enforce that
+    /// retargets only happen every 2016 blocks; instead, we assume that this
+    /// is the case if a valid proof of work is provided.
+    /// It is possible to cheat the relay by providing X blocks from earlier in
+    /// the most recent epoch, and then mining X new blocks after them.
+    /// However, each of these malicious blocks would have to be mined to a
+    /// higher difficulty than the legitimate ones.
```

Comment:
You can only reuse legitimate blocks that have been mined at the current difficulty D. Let's say the epoch began at time 10001 and ended (legitimately) at time 12016, and that the difficulty stays almost constant at ~D with this retarget. Let's say that blocks were mined every second during this period.

The last 6 blocks of the epoch were produced at 12011, 12012, 12013, 12014, 12015, and 12016.

If you try to reuse the blocks from this epoch, your last block at the given difficulty was produced either _at_ 12016 as in the legitimate chain, or earlier. If you present an earlier block as your last block, it results in a higher difficulty.

To abuse this, you'd need to use the blocks [12012 .. 12016] and then mine an additional block at time 18064, giving you the pre-retarget chain of [12012, 12013, 12014, 12015, 12016, 18064]. The addition of the malicious last block now means that our difficulty drops from D to D/4 because it looks like the epoch lasted 4 times the expected time.

If we add validation to the retarget so the timestamp of the last block of the epoch may not be in the future, this would give the legitimate chain plenty of time to pre-empt attempts to mine such chains (because such a difficulty-reducing chain could only be submitted significantly later than the real chain).